db.go

```diff
@@ -1836,6 +1836,113 @@ func (db *DB) dropAll() (func(), error) {
 	return resume, nil
 }
 
+// DropPrefixNonBlocking would logically drop all the keys with the provided prefix. The data would
+// not be cleared from LSM tree immediately. It would be deleted eventually through compactions.
+// This operation is useful when we don't want to block writes while we delete the prefixes.
+// It does this in the following way:
+// - Stream the given prefixes at a given ts.
+// - Write them to skiplist at the specified ts and handover that skiplist to DB.
+func (db *DB) DropPrefixNonBlocking(ts uint64, prefixes ...[]byte) error {
+	if db.opt.ReadOnly {
+		return errors.New("Attempting to drop data in read-only mode.")
+	}
+
+	if len(prefixes) == 0 {
+		return nil
+	}
+	db.opt.Infof("Non-blocking DropPrefix called for %s", prefixes)
+
+	dropPrefix := func(prefix []byte) error {
+		stream := db.NewStreamAt(ts)
+		stream.LogPrefix = fmt.Sprintf("Dropping prefix: %#x", prefix)
+		stream.Prefix = prefix
+		// Use the default implementation with some changes. We don't need anything except key.
+		stream.KeyToList = func(key []byte, itr *Iterator) (*pb.KVList, error) {
+			a := itr.Alloc
+			ka := a.Copy(key)
+
+			list := &pb.KVList{}
+			for ; itr.Valid(); itr.Next() {
+				item := itr.Item()
+				if item.IsDeletedOrExpired() {
+					break
+				}
+				if !bytes.Equal(key, item.Key()) {
+					// Break out on the first encounter with another key.
+					break
+				}
+
+				kv := y.NewKV(a)
+				kv.Key = ka
+				list.Kv = append(list.Kv, kv)
+
+				if db.opt.NumVersionsToKeep == 1 {
+					break
+				}
+
+				if item.DiscardEarlierVersions() {
+					break
+				}
+			}
+			return list, nil
+		}
+
+		var wg sync.WaitGroup
+		builderMap := make(map[uint32]*skl.Builder)
+		initSize := int64(float64(db.opt.MemTableSize) * 1.1)
+
+		handover := func(force bool) error {
+			for id, b := range builderMap {
+				sl := b.Skiplist()
+				if force || sl.MemSize() > db.opt.MemTableSize {
+					wg.Add(1)
+					if err := db.HandoverSkiplist(sl, wg.Done); err != nil {
+						return err
+					}
+					// Create a fresh builder.
+					builderMap[id] = skl.NewBuilder(initSize)
+				}
+			}
+			return nil
+		}
+
+		stream.Send = func(buf *z.Buffer) error {
+			err := buf.SliceIterate(func(s []byte) error {
+				var kv pb.KV
+				if err := kv.Unmarshal(s); err != nil {
+					return err
+				}
+				if _, ok := builderMap[kv.StreamId]; !ok {
+					builderMap[kv.StreamId] = skl.NewBuilder(initSize)
+				}
+				builderMap[kv.StreamId].Add(y.KeyWithTs(kv.Key, ts), y.ValueStruct{Meta: bitDelete})
+				return nil
+			})
+			if err != nil {
+				return err
+			}
+			return handover(false)
+		}
+		if err := stream.Orchestrate(context.Background()); err != nil {
+			return err
+		}
+		// Flush the remaining skiplists if any.
+		if err := handover(true); err != nil {
+			return err
+		}
+		wg.Wait()
```

Comment:
Is it okay to keep on waiting until the skiplist goes into the disk? Or should there be some kind of callback by the caller of `DropPrefixNonBlocking`?

Reply:
I think it should be okay. Because once the skiplist is pushed to `flushCh`, it should be flushed in some time. Unless the writes are stalled, which anyways is a problem.